### PR TITLE
Add basic check that the correct number of tests is built

### LINKED ISSUE #19286: Need automated mechanism to ensure we don't unintentionally lose test coverage
E.g., PR https://github.com/dotnet/coreclr/pull/19254 caused us to silently stop building pri-1 tests. This is silent because when we run tests, we run whatever tests we find.

One exception is Windows arm32/arm64, which currently have a separate list of exactly which tests we intend to run, and if the Pri-1 tests are there, they fail.

We need a mechanism that makes this a non-silent failure in all cases.

### PR DESCRIPTION
**Pull request by BruceForstall**
Fixes #19286

### CODE REVIEW

**Comment by BruceForstall** [comment]
@dotnet-bot test Windows_NT x64 Build and Test
@dotnet-bot test Windows_NT x86 Checked Build and Test

**Comment by BruceForstall** [comment]
@dotnet-bot test Windows_NT x86 Checked Build and Test

**Comment by BruceForstall** [comment]
@dotnet-bot test OSX10.12 x64 Checked CoreFX Tests

**Comment by jashook** [comment]
I understand the need for this type of check, it is something that was manually required to verify before #18695 was merged. As it is I suggest changes to this PR, either to add this information in one place in runtest.proj (as awful as that sounds it at least unifies it), or duplicate the logic in build-test.sh.

Third option is to wait for a build-test unification similar to runtest unification done in #19213

**Comment by BruceForstall** [comment]
@dotnet-bot test Windows_NT x64 Build and Test
@dotnet-bot test Windows_NT x86 Checked Build and Test

**Comment by BruceForstall** [comment]
@jashook How's this look?

**Comment by jashook** [comment]
Just needs the call from build-test.sh as well thanks!

**Comment by BruceForstall** [comment]
@dotnet-bot test Ubuntu x64 Build and Test
@dotnet-bot test Windows_NT x64 Build and Test
@dotnet-bot test Windows_NT x86 Checked Build and Test

File: build-test.cmd [hunk 1]
```diff
@@ -314,6 +314,10 @@ for /l %%G in (1, 1, %__BuildLoopCount%) do (
     set __AppendToLog=true
 )
 
+REM Check that we've built about as many tests as we expect. This is primarily intended to prevent accidental changes that cause us to build
+REM drastically fewer Pri-1 tests than expected.
+call %__DotnetHost% msbuild %__ProjectDir%\tests\runtest.proj /t:CheckTestBuild /p:Priority=%__Priority% %__msbuildArgs% %__unprocessedBuildArgs%
```
[the remark this answers is not in the document]
Comment (AaronRobinsonMSFT) [reply build-test.cmd]: Ugh.. Sorry, you are using `Priority` in your target.... instead, can you please use the standard property for tests: `CLRTestPriorityToBuild`

[the remark this answers is not in the document]
Comment (AaronRobinsonMSFT) [reply build-test.cmd]: BTW that is the core issue for why the build is failing....

[the remark this answers is not in the document]
Comment (BruceForstall) [reply build-test.cmd]: @AaronRobinsonMSFT I'm not invoking run.exe, I'm using msbuild directly.

**Comment by BruceForstall** [comment]
@dotnet-bot test Ubuntu x64 Build and Test
@dotnet-bot test Windows_NT x64 Build and Test
@dotnet-bot test Windows_NT x86 Checked Build and Test

**Comment by AaronRobinsonMSFT** [comment]
@BruceForstall I am generally against these kind of checks. This is the exact kind of code that atrophies within a week and then inevitably gets abused and confuses future team members, however I can appreciate the concern about loss of coverage - a real issue as demonstrated by the recent issue. What is  the long term maintenance approach for this code? At the very least it should be documented in one of the test markdown files so people know this mechanism exists.

**Comment by BruceForstall** [comment]
@dotnet-bot test Ubuntu x64 Build and Test
@dotnet-bot test Windows_NT x64 Build and Test
@dotnet-bot test Windows_NT x86 Checked Build and Test

**Comment by BruceForstall** [comment]
> @BruceForstall I am generally against these kind of checks. This is the exact kind of code that atrophies within a week and then inevitably gets abused and confuses future team members, however I can appreciate the concern about loss of coverage - a real issue as demonstrated by the recent issue. What is the long term maintenance approach for this code? At the very least it should be documented in one of the test markdown files so people know this mechanism exists and should be considered when working with tests in coreclr.

@AaronRobinsonMSFT I don't see how this code atrophies: if it fails, it will be a noisy failure. I don't see this as a maintenance issue: if it fails, because someone maybe added or removed a lot of tests, they'll see it. I've added a bunch of comments to the msbuild target. Note that the coverage lost last week was by no means the first time we've had this problem.

**Comment by AaronRobinsonMSFT** [comment]
This code atrophies because when more tests are added and the the count becomes 20000, when 8000 go missing it is again an issue. I am not saying there is no value, but any hard-coded value that represents a point in time quantity by definition atrophies if the quantity can change over time.

Adding comments in MSBuild isn't where people look for how to integrate into the system, it is where they look when something breaks. Knowing ahead of time that this mechanism exists is helpful so if someone adds many new tests or removes many, there is a heads up in the process document. Basically, this is exactly the kind of thing that should be mentioned in the documentation for working with tests in coreclr.

**Comment by Maoni0** [comment]
I just got this:

BUILDTEST: Check the managed tests build
D:\coreclr\tests\runtest.proj(75,5): error : Unexpected test count. Expected < 3000, found 7242.'

D:\coreclr\tests\runtest.proj(75,5): error : Unexpected test count. Expected < 3000, found 7242.'
BUILDTEST: Error: build failed.

**Pull request merged**